src/main/kotlin/com/terraformation/backend/accelerator/db/ApplicationStore.kt

```diff
@@ -0,0 +1,266 @@
+
+    val existing = fetchOneById(applicationId)
+    val modified = updateFunc(existing)
+
+    with(APPLICATIONS) {
+      dslContext
+          .update(APPLICATIONS)
+          .set(APPLICATION_STATUS_ID, modified.status)
+          .set(FEEDBACK, modified.feedback)
+          .set(INTERNAL_COMMENT, modified.internalComment)
+          .set(MODIFIED_BY, currentUser().userId)
+          .set(MODIFIED_TIME, clock.instant())
+          .where(ID.eq(applicationId))
+          .execute()
+    }
+
+    insertHistory(applicationId)
+  }
+
+  fun restart(applicationId: ApplicationId) {
+    requirePermissions { updateApplication(applicationId) }
+
+    val existing = fetchOneById(applicationId)
+
+    if (existing.status != ApplicationStatus.NotSubmitted) {
+      updateStatus(applicationId, ApplicationStatus.NotSubmitted)
+    }
+  }
+
+  fun submit(applicationId: ApplicationId) {
+    requirePermissions { updateApplication(applicationId) }
+
+    val existing = fetchOneById(applicationId)
+
+    if (existing.status == ApplicationStatus.NotSubmitted) {
+      updateStatus(applicationId, ApplicationStatus.Submitted)
+    } else {
+      log.info(
+          "Application $applicationId has status ${existing.status}; ignoring submission request")
+    }
+  }
+
+  private fun updateStatus(applicationId: ApplicationId, status: ApplicationStatus) {
+    with(APPLICATIONS) {
+      dslContext
+          .update(APPLICATIONS)
+          .set(APPLICATION_STATUS_ID, status)
+          .set(MODIFIED_BY, currentUser().userId)
+          .set(MODIFIED_TIME, clock.instant())
+          .where(ID.eq(applicationId))
+          .execute()
+    }
+
+    insertHistory(applicationId)
+  }
+
+  /**
+   * Updates an application's project boundary. If the application doesn't yet have an internal
+   * name, and the boundary is within a single country, sets the internal name based on the country
+   * and organization name.
+   */
+  fun updateBoundary(applicationId: ApplicationId, boundary: Geometry) {
+    requirePermissions { updateApplication(applicationId) }
+
+    dslContext.transaction { _ ->
+      val existing = fetchOneById(applicationId)
+
+      if (existing.internalName == null) {
+        val countries = countryDetector.getCountries(boundary)
+
+        if (countries.size == 1) {
+          val alpha3CountryCode =
+              countriesDao.fetchOneByCode(countries.single())?.codeAlpha3 ?: "XXX"
+          val organizationName =
+              organizationsDao.fetchOneById(existing.organizationId)?.name
+                  ?: throw OrganizationNotFoundException(existing.organizationId)
+          val internalName = "${alpha3CountryCode}_$organizationName"
+
+          updateInternalName(applicationId, internalName)
+        } else {
+          log.debug(
+              "Not setting internal name for application $applicationId because boundary is not " +
+                  "all in one country: $countries")
+        }
+      }
+
```

Comment:
should we validate the boundary size constraint here as well?

Reply:
I just posted a question on the PRD about this. The size checking behavior is a little inconsistent as specified.